src/Data/Set/NonEmpty.purs

```diff
@@ -0,0 +1,149 @@
+toUnfoldable (NonEmptySet s) = Set.toUnfoldable s
+
+-- | Convert a set to a non-empty unfoldable structure.
+toUnfoldable1 :: forall f a. Unfoldable1 f => NonEmptySet a -> f a
+toUnfoldable1 (NonEmptySet s) = unfoldr1 go (Set.toUnfoldable s :: List a)
+  where
+    go = unsafePartial case _ of
+      x : List.Nil -> Tuple x Nothing
+      x : tail -> Tuple x (Just tail)
+
+derive newtype instance eqNonEmptySet :: Eq a => Eq (NonEmptySet a)
+derive newtype instance eq1NonEmptySet :: Eq1 NonEmptySet
+derive newtype instance ordNonEmptySet :: Ord a => Ord (NonEmptySet a)
+derive newtype instance ord1NonEmptySet :: Ord1 NonEmptySet
+derive newtype instance semigroupNonEmptySet :: Ord a => Semigroup (NonEmptySet a)
+derive newtype instance foldableNonEmptySet :: Foldable NonEmptySet
+
+instance foldable1NonEmptySet :: Foldable1 NonEmptySet where
+  foldMap1 f = foldMap1 f <<< toUnfoldable1 :: forall a. NonEmptySet a -> NonEmptyList a
+  fold1 = foldMap1 identity
+
+instance showNonEmptySet :: Show a => Show (NonEmptySet a) where
+  show s = "(fromFoldable " <> show (toUnfoldable1 s :: NonEmptyList a) <> ")"
+
+-- | Create a set with one element.
+singleton :: forall a. a -> NonEmptySet a
+singleton a = NonEmptySet (Set.singleton a)
+
+-- | Maps over the values in a set.
+-- |
+-- | This operation is not structure-preserving for sets, so is not a valid
+-- | `Functor`. An example case: mapping `const x` over a set with `n > 0`
+-- | elements will result in a set with one element.
+map :: forall a b. Ord b => (a -> b) -> NonEmptySet a -> NonEmptySet b
+map f (NonEmptySet s) = NonEmptySet (Set.map f s)
+
+-- | Test if a value is a member of a set.
+member :: forall a. Ord a => a -> NonEmptySet a -> Boolean
+member a (NonEmptySet m) = Set.member a m
+
+-- | Insert a value into a set.
+insert :: forall a. Ord a => a -> NonEmptySet a -> NonEmptySet a
+insert a (NonEmptySet s) = NonEmptySet (Set.insert a s)
+
+-- | Delete a value from a non-empty set. If this would empty the set, the
+-- | result is `Nothing`.
+delete :: forall a. Ord a => a -> NonEmptySet a -> Maybe (NonEmptySet a)
+delete a (NonEmptySet s) = fromSet (Set.delete a s)
+
+-- | Find the size of a set.
+size :: forall a. NonEmptySet a -> Int
+size (NonEmptySet s) = Set.size s
+
+-- | The minimum value in the set.
+findMin :: forall a. NonEmptySet a -> a
+findMin (NonEmptySet s) = unsafePartial (fromJust (Set.findMin s))
+
+-- | The maximum value in the set.
+findMax :: forall a. NonEmptySet a -> a
+findMax (NonEmptySet s) = unsafePartial (fromJust (Set.findMax s))
+
+-- | Form the union of a set and the non-empty set.
+unionSet :: forall a. Ord a => Set.Set a -> NonEmptySet a -> NonEmptySet a
+unionSet s1 (NonEmptySet s2) = NonEmptySet (s1 <> s2)
+
+-- | Form the union of a non-empty collection of non-empty sets.
+unions :: forall f a. Foldable1 f => Ord a => f (NonEmptySet a) -> NonEmptySet a
+unions = foldl append (NonEmptySet Set.empty)
```

Comment:
Why not `unions = fold1`?

Reply:
Good thinking. Another one I thought of earlier today is that we can have a `Bounded` instance rather than providing `findMin` and `findMax` too.

Reply:
Yeah I don't know what I was thinking either 😁